### DISCUSSION TITLE
GRIB2_ipdtmpl_index and GRIB2_ipdtmpl_value usage

### DISCUSSION THREAD

**Discussion by BinbinZhou-NOAA** [dtcenter/METplus#1046] [Configuration] [answered]
I am verifying ensemble probabilitic 24hr precip forecast (output from ensemble_stat) by grid_stat
with probabilistic clim 24hr precip GRIB1 data as reference. My issue is all APCP IDs in this reference file are exactly same. Although the data sequence order is fixed according to the fixed thresholds, GRIB1 has difficult to handle different probability for different thresholds.  So I have to convert it to GRIB2 using specific table (Product template 4.5 )  which can handle thresholds.  The conversion is
fine. wgrib2 show following message now

1:0:d=2004072512:APCP:surface:anl:prob >-0.001
2:8363:d=2004072512:APCP:surface:anl:prob >0
3:17904:d=2004072512:APCP:surface:anl:prob >1
4:27363:d=2004072512:APCP:surface:anl:prob >2
5:36752:d=2004072512:APCP:surface:anl:prob >5
6:45993:d=2004072512:APCP:surface:anl:prob >10
7:55060:d=2004072512:APCP:surface:anl:prob >20
8:63893:d=2004072512:APCP:surface:anl:prob >25
9:72641:d=2004072512:APCP:surface:anl:prob >35
10:81276:d=2004072512:APCP:surface:anl:prob >50

 the product id template arrary for each threshold is like this
 
 PRODUCT TEMPLATE 4. 5:    1 8 2 0 184 0 0 1 0 1 0 0 255 0 0 0 0 1 0 0 3 1000 
The last (22th) "1000" is threshold value ( factored by 1000, that is >1 threshold) 
The arrays of all field probability fields are same except for this last value

Now in my grid_stat config file, fcst, obs and climo entries  like below
fcst = {

   level = "(*,*)";
   cat_thresh = [ ==0.1 ];
   prob = TRUE;

   field = [
      { name="APCP_24_A24_ENS_FREQ_gt-0.001"; },
      { name="APCP_24_A24_ENS_FREQ_gt0";    },
      { name="APCP_24_A24_ENS_FREQ_gt1";    },
      { name="APCP_24_A24_ENS_FREQ_gt2";    },
      { name="APCP_24_A24_ENS_FREQ_gt5";    },
      { name="APCP_24_A24_ENS_FREQ_gt10";   },
      { name="APCP_24_A24_ENS_FREQ_gt20";   },
      { name="APCP_24_A24_ENS_FREQ_gt25";   },
      { name="APCP_24_A24_ENS_FREQ_gt35";   },
      { name="APCP_24_A24_ENS_FREQ_gt50";   }
   ];

}

obs = {

   name  = "APCP_24";
   level = "A24";

   field = [
      { cat_thresh = [ gt-0.001 ];  },
      { cat_thresh = [ gt0  ];  },
      { cat_thresh = [ gt1  ];  },
      { cat_thresh = [ gt2  ];  },
      { cat_thresh = [ gt5  ];  },
      { cat_thresh = [ gt10 ];  },
      { cat_thresh = [ gt20 ];  },
      { cat_thresh = [ gt25 ];  },
      { cat_thresh = [ gt30 ];  },
      { cat_thresh = [ gt50 ];  }
   ];
}

climo_mean = fcst;
climo_mean = {

   file_name = "prob_24hr_20040726.grib2.prob";
   name  = "APCP";
   level = "*,*";

   field = [

         { GRIB2_ipdtmpl_index[ 21 ]; GRIB2_ipdtmpl_val=[ -1 ]; },
         { GRIB2_ipdtmpl_index[ 21 ]; GRIB2_ipdtmpl_val=[ 0 ]; },
         { GRIB2_ipdtmpl_index[ 21 ]; GRIB2_ipdtmpl_val=[ 1000 ]; },
         { GRIB2_ipdtmpl_index[ 21 ]; GRIB2_ipdtmpl_val=[ 2000 ]; },
         { GRIB2_ipdtmpl_index[ 21 ]; GRIB2_ipdtmpl_val=[ 5000 ]; },
         { GRIB2_ipdtmpl_index[ 21 ]; GRIB2_ipdtmpl_val=[ 10000 ]; },
         { GRIB2_ipdtmpl_index[ 21 ]; GRIB2_ipdtmpl_val=[ 20000 ]; },
         { GRIB2_ipdtmpl_index[ 21 ]; GRIB2_ipdtmpl_val=[ 25000 ]; },
         { GRIB2_ipdtmpl_index[ 21 ]; GRIB2_ipdtmpl_val=[ 35000 ]; },
         { GRIB2_ipdtmpl_index[ 21 ]; GRIB2_ipdtmpl_val=[ 50000 ]; }

   ];

   time_interp_method = DW_MEAN;
   match_day          = FALSE;

}

But running in above stting got error messge, says:
ERROR  :    text   = "["
ERROR  :
ERROR  :
ERROR  :          { GRIB2_ipdtmpl_index[ 21 ]; GRIB2_ipdtmpl_val=[ -1 ]; },
ERROR  : ___________________^______________________________________________
ERROR  :

It seems the error message means there is typo error, but I have no idea about this typo since 
the MET document suggests using these 2 entries to specify any one indexed value in 
the product ID template array

The entire testing directory is on WCOSS/venus
/gpfs/dell2/emc/modeling/noscrub/Binbin.Zhou/met/gefs_apcp24 
The running script is run.sh

Since this is just a testing case,   I don't use METplus  but  MET to simplify this case.

Binbin 
ECM/NCEP

**Comment by JohnHalleyGotway** [comment, marked as answer]
Looking carefully, I think I see the issue. There's an example of setting this in this [Chapter](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html?highlight=GRIB2_ipdtmpl_val#settings-common-to-multiple-tools) of the MET User's Guide: 
```
GRIB2_ipdtmpl_index=[8, 26];
GRIB2_ipdtmpl_val=[1, 2];
```
Comparing that to what you sent, please try switching to the following:
```
     { GRIB2_ipdtmpl_index=[ 21 ]; GRIB2_ipdtmpl_val=[ -1 ]; },
     { GRIB2_ipdtmpl_index=[ 21 ]; GRIB2_ipdtmpl_val=[ 0 ]; },
     { GRIB2_ipdtmpl_index=[ 21 ]; GRIB2_ipdtmpl_val=[ 1000 ]; },
     { GRIB2_ipdtmpl_index=[ 21 ]; GRIB2_ipdtmpl_val=[ 2000 ]; },
     { GRIB2_ipdtmpl_index=[ 21 ]; GRIB2_ipdtmpl_val=[ 5000 ]; },
     { GRIB2_ipdtmpl_index=[ 21 ]; GRIB2_ipdtmpl_val=[ 10000 ]; },
     { GRIB2_ipdtmpl_index=[ 21 ]; GRIB2_ipdtmpl_val=[ 20000 ]; },
     { GRIB2_ipdtmpl_index=[ 21 ]; GRIB2_ipdtmpl_val=[ 25000 ]; },
     { GRIB2_ipdtmpl_index=[ 21 ]; GRIB2_ipdtmpl_val=[ 35000 ]; },
     { GRIB2_ipdtmpl_index=[ 21 ]; GRIB2_ipdtmpl_val=[ 50000 ]; }
```

**Reply by BinbinZhou-NOAA** [reply]
After this change, it works now.

Thanks!

Binbin
